pymc_marketing/mmm/evaluation.py

```diff
@@ -0,0 +1,221 @@
+        logger.error(f"Model has {divergences} divergences")
+    else:
+        logger.success("No divergences!")
+
+    # Calculate elemwise log_likelihood of model given posteriors
+    pm.compute_log_likelihood(
+        mmm.idata, model=mmm.model, progressbar=False
+    )  # Used for LOOCV
+    model_loo = az.loo(mmm.idata)
+
+    # Log LOOCV metrics
+    loocv_metrics = {
+        "loocv_elpd_loo": model_loo.elpd_loo,  # expected log pointwise predictive density
+        "loocv_se": model_loo.se,  # standard error of elpd
+        "loocv_p_loo": model_loo.p_loo,  # effective number of parameters
+    }
+
+    # Display LOOCV metrics (including the Pareto k values which aren't logged in MLFlow)
+    logger.info(f"LOOCV metrics: {model_loo}")
+    # Combine diagnostics and LOOCV metrics
+    model_diagnostics.update(loocv_metrics)
+
+    return model_diagnostics, model_loo
+
+
+def plot_hdi_forest(mmm: MMM, var_names: list) -> plt.Figure:
+    """
+    Plot a forest plot to compare high-density intervals (HDIs) from a given set of
+    posterior distributions, as well as their r-hat statistic.
+
+    Parameters:
+    - mmm: Model object with inference data
+    - var_names: list: List of variable names to include in the forest plot
+
+    Returns:
+    - plt.Figure: The matplotlib figure object
+    """
+    # Ensure that mmm has the required 'idata' attribute and is not None
+    if not hasattr(mmm, "idata") or mmm.idata is None:
+        raise ValueError(
+            "The model object 'mmm' must have a non-None 'idata' attribute. "
+            "Ensure you've called mmm.fit() before running diagnostics."
+        )
+    # Create the forest plot
+    fig, ax = plt.subplots(figsize=(12, 8), ncols=2)
+    az.plot_forest(
+        data=mmm.idata,
+        var_names=var_names,
+        combined=True,
+        ax=ax,
+        hdi_prob=0.95,
+        # Also plot the split R-hat statistic
+        r_hat=True,
+    )
+
+    # Set the title for the figure
+    fig.suptitle("Posterior Distributions: 95.0% HDI")
+
+    # Return the figure
+    return fig
+
+
+def calc_metrics(
+    y_true: np.ndarray, y_pred: np.ndarray, prefix: str | None = None
+) -> dict[str, float]:
+    """
+    Calculate metrics including R-squared, RMSE, NRMSE, MAE, NMAE, and MAPE for a given true and predicted dataset.
+
+    Parameters:
+    - y_true: np.ndarray: True values for the dataset.
+    - y_pred: np.ndarray: Predictions for the dataset.
+    - prefix: Optional[str]: Prefix to label the metrics (e.g., 'train', 'test'). Defaults to None.
+
+    Returns:
+    - Dict[str, float]: A dictionary containing calculated metrics.
+    """
+    if prefix is None:
+        prefix = ""
+    else:
+        prefix += "_"
+
+    metrics = {
+        f"{prefix}r_squared": np.nan,
+        f"{prefix}rmse": np.nan,
+        f"{prefix}nrmse": np.nan,
+        f"{prefix}mae": np.nan,
+        f"{prefix}nmae": np.nan,
+        f"{prefix}mape": np.nan,
+    }
+
+    try:
```

Comment:
What is a reason these would error? If one fails, would they all fail?

Reply:
Ah this was holdover code from an earlier version I made - I'll delete!